### PR TITLE
KNOX-2890 - Prevent non-idempotent requests from failing over

### PR DESCRIPTION
**Pull request by moresandeep**
## What changes were proposed in this pull request?

Make sure we don't failover for non-idempotent requests such as POST. The request will anyways fail because the input response stream has already been closed so there is no point "retrying" the request. 

Following are the proposed changes:

- Prevent failover for non-idempotent requests in `ConfigurableHADispatch`
- Prevent failover for non-idempotent requests in `AbstractHdfsHaDispatch`, unfortunately `AbstractHdfsHaDispatch` was not using `ConfigurableHADispatch`, with this change it will be extending `ConfigurableHADispatch`.
- Also added is a configuration option for HA Provider `failoverNonIdempotentRequestEnabled` which will enable/disable failover for non-idempotent requests. Default value of `failoverNonIdempotentRequestEnabled` is `false`

e.g.
```
<param>
     <name>WEBHDFS</name>                       
     <value>enabled=true;maxFailoverAttempts=3;failoverSleep=1000;failoverNonIdempotentRequestEnabled=true</value>
</param>
```

## How was this patch tested?
Tested locally, added unit tests.

### CODE REVIEW

**Review by pzampino — CHANGES_REQUESTED** [review]

File: gateway-provider-ha/src/main/java/org/apache/knox/gateway/ha/dispatch/ConfigurableHADispatch.java [hunk 1]
```diff
@@ -61,18 +61,20 @@ public class ConfigurableHADispatch extends ConfigurableDispatch {
...
 
-  private int failoverSleep = HaServiceConfigConstants.DEFAULT_FAILOVER_SLEEP;
+  protected int failoverSleep = HaServiceConfigConstants.DEFAULT_FAILOVER_SLEEP;
 
-  private HaProvider haProvider;
+  protected HaProvider haProvider;
 
   private static final Map<String, String> urlToHashLookup = new HashMap<>();
   private static final Map<String, String> hashToUrlLookup = new HashMap<>();
+  protected static final List<String> idempotentRequests = Arrays.asList("POST", "PATCH", "CONNECT");
```
[the remark this answers is not in the document]
Comment (moresandeep) [reply gateway-provider-ha/src/main/java/org/apache/knox/gateway/ha/dispatch/ConfigurableHADispatch.java:72]: darn it, naming, i'll change it. Good Catch!

File: gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java [hunk 2]
```diff
@@ -78,14 +60,35 @@ protected void executeRequest(HttpUriRequest outboundRequest, HttpServletRequest
          inboundResponse = executeOutboundRequest(outboundRequest);
          writeOutboundResponse(outboundRequest, inboundRequest, outboundResponse, inboundResponse);
       } catch (StandbyException e) {
-         LOG.errorReceivedFromStandbyNode(e);
-         failoverRequest(outboundRequest, inboundRequest, outboundResponse, inboundResponse, e);
+        /* if non-idempotent requests are not allowed to failover */
```
[the remark this answers is not in the document]
Comment (pzampino) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java:63]: I do now notice that the only difference is the log message in the failover cases.

[the remark this answers is not in the document]
Comment (pzampino) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java:63]: Could still be factored out to a common method though, I think.

[the remark this answers is not in the document]
Comment (moresandeep) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java:63]: Yup, I can squash the code to take out duplicates.

[the remark this answers is not in the document]
Comment (moresandeep) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java:63]: Oh you mean, like make it a method that does the if ..else logic for logging?

File: gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java [hunk 3]
```diff
@@ -59,33 +59,19 @@ protected void executeRequest(HttpUriRequest outboundRequest, HttpServletRequest
...
-              inboundResponse, e);
-        }
-      } catch (IOException e) {
-        /* if non-idempotent requests are not allowed to failover */
-        if(!failoverNonIdempotentRequestEnabled && idempotentRequests.stream().anyMatch(outboundRequest.getMethod()::equalsIgnoreCase)) {
-          LOG.cannotFailoverToNonIdempotentRequest(outboundRequest.getMethod());
-          throw e;
-        } else {
-          LOG.errorConnectingToServer(outboundRequest.getURI().toString(), e);
+          if(e instanceof StandbyException) {
```
Comment (pzampino) [inline gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java]: Did we lose the failoverRequest invocation here?

Comment (moresandeep) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java]: Nope, it's at the bottom, below the if else statement (new refactored into a method).

Comment (pzampino) [reply gateway-service-webhdfs/src/main/java/org/apache/knox/gateway/hdfs/dispatch/AbstractHdfsHaDispatch.java]: ok, must be the multiple commits confusing me.

**Comment by pzampino** [comment]
LGTM, thanks!

**Review by pzampino — COMMENTED** [review]
LGTM

**Pull request merged**